---
title: "qc metric analysis"
author: "rhea shah"
date: "2023-07-11"
output: html_document
---

```{r}

library(dplyr)
library(Seurat)
library(patchwork)
library(hdf5r)
library(ggplot2)
library(magrittr)

# ================= rhea - eyeballing based on distribution =================

# ----------------------- loading in the data -----------------------
# loading the data
hce048_obj <- Read10X_h5(filename = "~/2023 sir/single cell analysis/hce048_distal_filteredfeaturematrix.h5")

# creating the seurat object
hce048_obj <- CreateSeuratObject(counts = hce048_obj)

# importing the hce048 proximal data
hce048_proximal_obj <- Read10X_h5(filename = "~/2023 sir/single cell analysis/hce048_proximal_filteredfeaturematrix.h5")

# creating the seurat object
hce048_proximal_obj <- CreateSeuratObject(counts = hce048_proximal_obj)

# importing the iw0076 distal data
iw0076_distal_obj <- Read10X_h5(filename = "~/2023 sir/single cell analysis/iw0076_distal_filteredfeaturematrix.h5")

# creating the seurat object
iw0076_distal_obj <- CreateSeuratObject(counts = iw0076_distal_obj)

# importing the iw0076 proximal data
iw0076_proximal_obj <- Read10X_h5(filename = "~/2023 sir/single cell analysis/iw0076_proximal_filteredfeaturematrix.h5")

# creating the seurat object
iw0076_proximal_obj <- CreateSeuratObject(counts = iw0076_proximal_obj)

# importing the p0630 distal data
p0630_distal_obj <- Read10X_h5(filename = "~/2023 sir/single cell analysis/p0630_distal_filteredfeaturematrix.h5")

# creating the seurat object
p0630_distal_obj <- CreateSeuratObject(counts = p0630_distal_obj)

# importing the p0630 proximal data
p0630_proximal_obj <- Read10X_h5(filename = "~/2023 sir/single cell analysis/p0630_proximal_filteredfeaturematrix.h5")

# creating the seurat object
p0630_proximal_obj <- CreateSeuratObject(counts = p0630_proximal_obj)


# ------------------------- FOR LOOP ----------------------------------------

sample_list <- list(hce048_obj, hce048_proximal_obj, iw0076_distal_obj, iw0076_proximal_obj, p0630_distal_obj, p0630_proximal_obj)  

for (object in sample_list) {
   # adding number of genes per umi for each cell to the metadata
   object$log10GenesPerUmi <- log10(object$nFeature_RNA + 1) / log10(object$nCount_RNA + 1)  # Adding 1 to avoid division by zero
   
   # adding the percent.mt RATIO (why we divide by 100)
   object$mitoRatio <- PercentageFeatureSet(object = object, pattern = "^MT-")
   object$mitoRatio <- object$mitoRatio / 100  # Access the mitoRatio column directly
   
   # creating the meta data dataframe
   metadata <- object@meta.data  # Access the meta.data slot of the object
   
   # adding cell ids to the metadata
   metadata$cells <- rownames(metadata)
   
   # renaming some of the columns
   metadata <- metadata %>%
     rename(seq_folder = orig.ident,
            nUMI = nCount_RNA,
            nGene = nFeature_RNA)
   
   # creating another column to group the cells together to make it easier to visualize
   metadata$sample <- "cell"
   
   # adding the metadata back into the seurat object
   object@meta.data <- metadata
   
   # visualizing the number of cell counts
   cell_counts <- metadata %>%
     ggplot(aes(x = sample)) +
     geom_bar() +
     theme_classic() +
     theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
     theme(plot.title = element_text(hjust=0.5, face="bold")) +
     ggtitle("nCells")
   
   # visualizing umi's per cell
   umi_per_cell <- metadata %>%
     ggplot(aes(color=sample, x=nUMI, fill= sample)) +
     geom_density(alpha = 0.2) +
     scale_x_log10() +
     theme_classic() +
     ylab("Cell density") +
     geom_vline(xintercept = 50)
   
   # Visualize the distribution of genes detected per cell via histogram
   genes_per_cell_hist <- metadata %>%
     ggplot(aes(color=sample, x=nGene, fill= sample)) +
     geom_density(alpha = 0.2) +
     theme_classic() +
     scale_x_log10() +
     geom_vline(xintercept = 300)
   
   # Visualize the distribution of genes detected per cell via boxplot
   genes_per_cell_box <- metadata %>%
     ggplot(aes(x=sample, y=log10(nGene), fill=sample)) +
     geom_boxplot() +
     theme_classic() +
     theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
     theme(plot.title = element_text(hjust=0.5, face="bold")) +
     ggtitle("NCells vs NGenes")
   
   # Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
   gene_and_umi <- metadata %>%
     ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) +
     geom_point() +
     scale_colour_gradient(low = "gray90", high = "black") +
     stat_smooth(method=lm) +
     scale_x_log10() +
     scale_y_log10() +
     theme_classic() +
     geom_vline(xintercept = 500) +
     geom_hline(yintercept = 250) +
     facet_wrap(~sample)
   
   # Visualize the distribution of mitochondrial gene expression detected per cell
   mito_expression <- metadata %>%
     ggplot(aes(color=sample, x=mitoRatio, fill=sample)) +
     geom_density(alpha = 0.2) +
     scale_x_log10() +
     theme_classic() +
     geom_vline(xintercept = 0.2)
   
   # Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
   gene_per_umi <- metadata %>%
     ggplot(aes(x=log10GenesPerUmi, color = sample, fill=sample)) +
     geom_density(alpha = 0.2) +
     theme_classic() +
     geom_vline(xintercept = 0.8)
   
   #outputting the various graphs
   print(cell_counts)
   print(gene_per_umi)
   print(umi_per_cell)
   print(mito_expression)
   print(gene_and_umi)
   print(genes_per_cell_box)
   print(genes_per_cell_hist)
}

# --------------------------- creating the subsets ------------------------------

# creating a subset of the hce048 distal data
filtered_hce048sample <- subset(hce048_obj@meta.data, subset = (nGene >= 250) & (mitoRatio < 0.2) & (log10GenesPerUmi > 0.75) & (nUMI > 750))

# creating a subset of hce048 proximal data 
filtered_hce048proximalsample <- subset(hce048_proximal_obj, subset = (nUMI > 750) & (nGene >= 500) & (mitoRatio < 0.2) & (log10GenesPerUmi > 0.7))

# subset for the iw0076 distal data 
filtered_iw0076distalsample <- subset(iw0076_distal_obj, subset = (nUMI > 500) & (nGene >= 750) & (mitoRatio < 0.2) & (log10GenesPerUmi > 0.7))

# subset for the iw0076 proximal data 
filtered_iw0076proximalsample <- subset(iw0076_proximal_obj, subset = (nUMI > 500) & (nGene >= 500) & (mitoRatio < 0.2) & (log10GenesPerUmi > 0.7))

# subset for p0630 distal data
filtered_p0630distalsample <- subset(p0630_distal_obj, subset = (nUMI > 500) & (nGene >= 500) & (mitoRatio < 0.2) & (log10GenesPerUmi > 0.7))

# subset for p0630 proximal data 
filtered_p0630proximalsample <- subset(p0630_proximal_obj, subset = (nUMI > 750) & (nGene >= 300) & (mitoRatio < 0.2) & (log10GenesPerUmi > 0.75))


# ================= ankitha - relative to median =================
library(dplyr)
library(Seurat)
library(SeuratObject)
library(ggplot2)
library(hdf5r)

HCD.data <- Read10X("/Users/ankithachalla/Desktop/RStudio/SingleCellProject/OneDrive_2_7-10-2023/HCE048-D_MP_cellranger6_01182023/outs/filtered_feature_bc_matrix/")
HCD <- CreateSeuratObject(counts = HCD.data, project = "HCE048-D")

HCP.data <- Read10X("/Users/ankithachalla/Desktop/RStudio/SingleCellProject/OneDrive_2_7-10-2023/HCE048-P_MP_cellranger6_01182023/outs/filtered_feature_bc_matrix/")
HCP <- CreateSeuratObject(counts = HCP.data, project = "HCE048-P")

IWD.data <- Read10X("/Users/ankithachalla/Desktop/RStudio/SingleCellProject/OneDrive_1_7-10-2023/IW0076-D_MP_cellranger6_01182023/outs/filtered_feature_bc_matrix/")
IWD <- CreateSeuratObject(counts = IWD.data, project = "IW0076-D")

IWP.data <- Read10X("/Users/ankithachalla/Desktop/RStudio/SingleCellProject/OneDrive_1_7-10-2023/IW0076-P_MP_cellranger6_01182023/outs/filtered_feature_bc_matrix/")
IWP <- CreateSeuratObject(counts = IWP.data, project = "IW0076-P")

P0D.data <- Read10X("/Users/ankithachalla/Desktop/RStudio/SingleCellProject/P0630-D filtered_feature_bc_matrix/")
P0D <- CreateSeuratObject(counts = P0D.data, project = "P0630-D")

P0P.data <- Read10X_h5("/Users/ankithachalla/Desktop/RStudio/SingleCellProject/P0630-P filtered_feature_bc_matrix.h5")
PWP <- CreateSeuratObject(counts = P0P.data, project = "P0630-P")

# Calculating percentage of mitochondrial genes per cell
HCD[["percent.mt"]] <- PercentageFeatureSet(HCD, pattern = "^MT-")
HCP[["percent.mt"]] <- PercentageFeatureSet(HCP, pattern = "^MT-")
IWD[["percent.mt"]] <- PercentageFeatureSet(IWD, pattern = "^MT-")
IWP[["percent.mt"]] <- PercentageFeatureSet(IWP, pattern = "^MT-")
P0D[["percent.mt"]] <- PercentageFeatureSet(P0D, pattern = "^MT-")
P0P[["percent.mt"]] <- PercentageFeatureSet(P0P, pattern = "^MT-")

# Calculating percentage of ribosomal genes per cell
HCD[["percent.rb"]] <- PercentageFeatureSet(HCD, pattern = "^RP[SL]")
HCP[["percent.rb"]] <- PercentageFeatureSet(HCP, pattern = "^RP[SL]")
IWD[["percent.rb"]] <- PercentageFeatureSet(IWD, pattern = "^RP[SL]")
IWP[["percent.rb"]] <- PercentageFeatureSet(IWP, pattern = "^RP[SL]")
P0D[["percent.rb"]] <- PercentageFeatureSet(P0D, pattern = "^RP[SL]")
P0P[["percent.rb"]] <- PercentageFeatureSet(P0P, pattern = "^RP[SL]")

med_calc <- function(obj) {
  # Calculating median value of each metric
  count_med <- median(obj@meta.data$nCount_RNA)
  feature_med <- median(obj@meta.data$nFeature_RNA)
  pct_mt_med <- median(obj@meta.data$percent.mt)
  pct_rb_med <- median(obj@meta.data$percent.rb)
  
  return(c(count_med, feature_med, pct_mt_med, pct_rb_med))
}

mad_calc <- function(obj) {
    count_med <- med_calc(obj)[1]
    feature_med <- med_calc(obj)[2]
    pct_mt_med <- med_calc(obj)[3]
    pct_rb_med <- med_calc(obj)[4]
  
    # Calculates absolute deviation (subtract median from each value)
    nCt_abs_dev <- abs(obj@meta.data$nCount_RNA - count_med)
    nFt_abs_dev <- abs(obj@meta.data$nFeature_RNA - feature_med)
    mt_abs_dev <- abs(obj@meta.data$percent.mt - pct_mt_med)
    rb_abs_dev <- abs(obj@meta.data$percent.rb - pct_rb_med)


    # Calculates the median absolute deviation
    nCt_mad <- mad(nCt_abs_dev, constant = 1)
    nFt_mad <- mad(nFt_abs_dev, feature_med)
    mt_mad <- mad(mt_abs_dev, pct_mt_med)
    rb_mad <- mad(rb_abs_dev, pct_rb_med)
    
    return(c(nCt_mad, nFt_mad, mt_mad, rb_mad))
}

mad_filter <- function(obj) {
  count_med <- med_calc(obj)[1]
  feature_med <- med_calc(obj)[2]
  pct_mt_med <- med_calc(obj)[3]
  pct_rb_med <- med_calc(obj)[4]
  
  nCt_mad <- mad_calc(obj)[1]
  nFt_mad <- mad_calc(obj)[2]
  mt_mad <- mad_calc(obj)[3]
  rb_mad <- mad_calc(obj)[4]
  
  # Filtering Cells if they fall within the range
  obj@meta.data[["nCount_Cutoff"]] <- ifelse(obj@meta.data$nCount_RNA >= (count_med - (3*nCt_mad)) & obj@meta.data$nCount_RNA <= (count_med + (3*nCt_mad)), TRUE, FALSE)
  obj@meta.data[["nFeature_Cutoff"]] <- ifelse(obj@meta.data$nFeature_RNA >= (feature_med - (3*nFt_mad)) & obj@meta.data$nFeature_RNA <= (feature_med + (3*nFt_mad)), TRUE, FALSE)
  obj@meta.data[["percent.mt_Cutoff"]] <- ifelse(obj@meta.data$percent.mt >= (pct_mt_med - (3*mt_mad)) & obj@meta.data$percent.mt <= (pct_mt_med + (3*mt_mad)), TRUE, FALSE)
  obj@meta.data[["percent.rb_Cutoff"]] <- ifelse(obj@meta.data$percent.rb >= (pct_rb_med - (3*rb_mad)) & obj@meta.data$percent.rb <= (pct_rb_med + (3*rb_mad)), TRUE, FALSE)
  
  # Subsetting the cells
  obj_mad <- subset(obj, nCount_Cutoff == TRUE & nFeature_Cutoff == TRUE & percent.mt_Cutoff == TRUE & percent.rb_Cutoff == TRUE)
  
  return(c(obj, obj_mad))
}

# ================= himani - absolute value =================

library(dplyr)
library(Seurat)
library(patchwork)
library(hdf5r)

# Load the data
obj.data <- Read10X_h5("C:/Users/Himani/Downloads/matrix.h5")

# Create the seurat object
seurat_obj <- CreateSeuratObject(counts = obj.data, min.cells = 3, min.features = 200)

# visualizing the object before the QC strategy using a Vln Plot
seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, pattern = "^MT-")
VlnPlot(seurat_obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), pt.size = 0.01, ncol = 3)

# Scatterplot of total counts vs. number of expressed genes
FeatureScatter(seurat_obj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

#using absolute value QC strategy (identifying & removing low-quality cells)

# Define cutoff values
count_cutoff <- 4000
gene_cutoff <- 3000

# Filter cells based on cutoffs
seurat_obj <- subset(seurat_obj, subset = nCount_RNA > count_cutoff)

# Filter cells by gene
seurat_obj <- subset(seurat_obj, subset = nFeature_RNA > gene_cutoff)

#updating the seurat object
updated_seurat <- UpdateSeuratObject(seurat_obj)

#visualizing it after the changes
seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, pattern = "^MT-")
VlnPlot(seurat_obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), pt.size = 0.01, ncol = 3)

# ================= anjali - percentile =================  



```